### DISCUSSION TITLE
Slow inference speed

### DISCUSSION THREAD

**Discussion by lachioma** [talmolab/sleap#1099] [Help!]
Hello

I am trying to use Sleap for real-time tracking and I would need the fastest possible inference speed (latency around or <10 msec).
Using my Nvidia RTX 3090, I report an inference latency on single frames of 70-80 msec (480 x 300 pixels, 2 body parts, top-down configuration with 1 instance), which is quite slow compared to what reported on the [Nature Methods paper](https://www.nature.com/articles/s41592-022-01426-1#Sec9:~:text=Detection%20and%20control%20of%20behavior%20in%20real%20time) and by other users. 

So I really think that something is wrong with my system, but I do not quite get what. 
Any help please??

After reading [this discussion](https://github.com/talmolab/sleap/discussions/833), I tried different gpu driver versions (the latest driver 527.56; the one that comes together with Cuda 11.3, 465.89; one in the middle, released right before the release of TensorFlow 2.6.3 (the version sleap is based on) 511.23),  but it doesn't help.
I also tried installing TensorFlow 2.8 and 2.10 but it doesn't help either.

I am running Sleap with python, using `predictor.inference_model.predict_on_batch(image)`, as documented in the [notebook on realtime inference using SLEAP](https://sleap.ai/notebooks/Interactive_and_realtime_inference.html).

Here is the output of `sleap-diagnostic` for my current system:

```
(sleap) PS C:\Users\al6978> sleap-diagnostic
INFO:numexpr.utils:Note: NumExpr detected 36 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
INFO:numexpr.utils:NumExpr defaulting to 8 threads.

==========SYSTEM==========

utc:                    2022-12-30 20:48:40.968225
python:                 3.7.12
system:                 Windows, AMD64, 10, 10.0.19041
path:                   C:\Users\al6978\.conda\envs\sleap\lib\site-packages\PySide2;C:\Users\al6978\.conda\envs\sleap\lib\site-packages\cv2\../../x64/vc14/bin;C:\Users\al6978\.conda\envs\sleap\lib\site-packages\pywin32_system32;C:\Users\al6978\.conda\envs\sleap;C:\Users\al6978\.conda\envs\sleap\Library\mingw-w64\bin;C:\Users\al6978\.conda\envs\sleap\Library\usr\bin;C:\Users\al6978\.conda\envs\sleap\Library\bin;C:\Users\al6978\.conda\envs\sleap\Scripts;C:\Users\al6978\.conda\envs\sleap\bin;C:\ProgramData\Anaconda3\condabin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.7\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.7\libnvvp;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\dotnet;C:\Program Files (x86)\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\MATLAB\R2022b\bin;C:\Program Files\NVIDIA Corporation\Nsight Compute 2022.2.1;C:\Program Files\NVIDIA Corporation\NVIDIA NvDLISR;C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015;C:\Program Files\Point Grey Research\Spinnaker\bin\vs2015;C:\Program Files\Git\cmd;C:\Users\al6978\AppData\Local\Microsoft\WindowsApps;C:\Users\al6978\AppData\Local\GitHubDesktop\bin;C:\FFmpeg\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\zlib123dllx64\dll_x64;C:\Program Files\PsychoPy;C:\Program Files\PsychoPy\DLLs;C:\Users\al6978\.conda\envs\sleap\lib\site-packages\scipy\.libs

==========IMPORTS==========

sleap import:                   True
sleap path:                     C:\Users\al6978\.conda\envs\sleap\lib\site-packages\sleap\__init__.py
sleap version:                  1.2.9
pyside2 import:                 True
pyside path:                    C:\Users\al6978\.conda\envs\sleap\lib\site-packages\PySide2\__init__.py
call to C:\Users\al6978\.conda\envs\sleap\python.exe failed
C:\Users\al6978\.conda\envs\sleap\python.exe: can't open file 'C:\Users\al6978\.conda\envs\sleap\Scripts\sleap-diagnostic': [Errno 2] No such file or directory

pyside6 import:                 False
cv2 import:                     True

==========GIT==========

call to git failed
fatal: not a git repository (or any of the parent directories): .git

call to git failed
fatal: not a git repository (or any of the parent directories): .git


==========TENSORFLOW==========

tensorflow import:                      True
tensorflow version:                     2.6.3
tensorflow path:                        C:\Users\al6978\AppData\Roaming\Python\Python37\site-packages\tensorflow\__init__.py
gpus:                   [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

==========CONDA==========

unable to locate conda

==========PIP==========

absl-py==0.15.0
astunparse==1.6.3
attrs==21.2.0
backcall==0.2.0
backports.zoneinfo==0.2.1
Bottleneck @ file:///D:/bld/bottleneck_1656803876399/work
cached-property @ file:///home/conda/feedstock_root/build_artifacts/cached_property_1615209429212/work
cachetools==4.2.4
cattrs==1.1.1
certifi==2021.10.8
charset-normalizer==2.0.12
clang==5.0
cloudpickle==2.2.0
colorama==0.4.6
commonmark==0.9.1
cycler==0.11.0
debugpy==1.6.4
decorator==5.1.1
efficientnet==1.0.0
entrypoints==0.4
flatbuffers==1.12
fonttools==4.38.0
gast==0.4.0
google-auth==1.35.0
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
grpcio==1.44.0
h5py @ file:///D:/bld/h5py_1604753757907/work
hdmf==3.4.7
idna==3.3
image-classifiers==1.0.0
imageio==2.15.0
imgaug==0.4.0
imgstore==0.2.9
importlib-metadata==4.11.1
importlib-resources==5.10.0
ipykernel==6.16.2
ipython==7.34.0
jedi==0.18.2
joblib==1.2.0
jsmin==3.0.1
jsonpickle==1.2
jsonschema==4.17.0
jupyter_client==7.4.8
jupyter_core==4.12.0
keras==2.6.0
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.2
kiwisolver==1.4.4
Markdown==3.3.6
matplotlib==3.5.3
matplotlib-inline==0.1.6
ndx-pose==0.1.1
nest-asyncio==1.5.6
networkx==2.6.3
numexpr @ file:///D:/bld/numexpr_1637086525232/work
numpy @ file:///D:/bld/numpy_1649281563521/work
oauthlib==3.2.0
olefile @ file:///home/conda/feedstock_root/build_artifacts/olefile_1602866521163/work
opencv-python @ git+https://github.com/talmolab/wrap_opencv-python-headless.git@ede49f6a23a73033216339f29515e59d594ba921
opencv-python-headless==4.5.5.62
opt-einsum==3.3.0
packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1670530880680/work
pandas @ file:///C:/ci/pandas_1641461312795/work
parso==0.8.3
pickleshare==0.7.5
Pillow==8.4.0
pkgutil_resolve_name==1.3.10
prompt-toolkit==3.0.36
protobuf==3.19.4
psutil==5.9.4
pyasn1==0.4.8
pyasn1-modules==0.2.8
Pygments==2.13.0
pykalman==0.9.5
pynwb==2.2.0
pyparsing==3.0.7
pyreadline @ file:///D:/bld/pyreadline_1655703530615/work
pyrsistent==0.19.2
PySide2==5.14.1
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/python-dateutil_1626286286081/work
python-rapidjson==1.9
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1671365381334/work
pytz-deprecation-shim==0.1.0.post0
PyWavelets==1.3.0
pywin32==305
PyYAML==6.0
pyzmq==24.0.1
qimage2ndarray==1.9.0
QtPy @ file:///home/conda/feedstock_root/build_artifacts/qtpy_1667873092748/work
requests==2.27.1
requests-oauthlib==1.3.1
rich==10.16.1
rsa==4.8
ruamel.yaml==0.17.21
ruamel.yaml.clib==0.2.7
scikit-image==0.19.3
scikit-learn==1.0.2
scikit-video==1.1.11
scipy @ file:///C:/bld/scipy_1637806996411/work
seaborn==0.12.1
segmentation-models==1.0.1
setuptools-scm==6.4.2
Shapely==1.7.1
shiboken2==5.14.1
six @ file:///home/conda/feedstock_root/build_artifacts/six_1590081179328/work
sleap==1.2.9
spyder-kernels==2.3.3
tensorboard==2.6.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow==2.6.3
tensorflow-estimator==2.6.0
termcolor==1.1.0
threadpoolctl==3.1.0
tifffile==2021.11.2
tomli==2.0.1
tornado==6.2
traitlets==5.8.0
typing-extensions==3.10.0.2
tzdata==2022.6
tzlocal==4.2
urllib3==1.26.8
wcwidth==0.2.5
Werkzeug==2.0.3
wrapt==1.12.1
zipp==3.7.0


==========NVIDIA==========

Fri Dec 30 15:48:43 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 511.23       Driver Version: 511.23       CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:17:00.0  On |                  N/A |
|  0%   39C    P8    33W / 350W |   2657MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1872    C+G                                   N/A      |
|    0   N/A  N/A      5444    C+G   ...me\Application\chrome.exe    N/A      |
|    0   N/A  N/A      7960    C+G   C:\Windows\explorer.exe         N/A      |
|    0   N/A  N/A      9932    C+G   ...lPanel\SystemSettings.exe    N/A      |
|    0   N/A  N/A     10124    C+G   ...eripheral Manager\DPM.exe    N/A      |
|    0   N/A  N/A     11420    C+G   ...artMenuExperienceHost.exe    N/A      |
|    0   N/A  N/A     11824    C+G   ...5n1h2txyewy\SearchApp.exe    N/A      |
|    0   N/A  N/A     12952    C+G   ...e\PhoneExperienceHost.exe    N/A      |
|    0   N/A  N/A     14020    C+G   ...2txyewy\TextInputHost.exe    N/A      |
|    0   N/A  N/A     15692    C+G   ...ge\Application\msedge.exe    N/A      |
|    0   N/A  N/A     17000    C+G   ...ck\app-4.29.149\slack.exe    N/A      |
|    0   N/A  N/A     18280    C+G   ...3d8bbwe\CalculatorApp.exe    N/A      |
|    0   N/A  N/A     22952      C   ...nda\envs\sleap\python.exe    N/A      |
+-----------------------------------------------------------------------------+

```

**Comment by talmo** [comment]
Hi @lachioma,

Thanks for providing the detailed description of the setup!

70-80 ms on a 3090 is definitely way too high for 480 x 300 with only 2 nodes. Our baseline with 1024 x 1024 images, 13 nodes, top-down (with 2x crops of 160 x 160) is roughly 10-12 ms with batch size of 1 with pure TensorFlow on an TItan RTX.

Do you mind sharing the `training_job_config.json` for the model(s) you've been benchmarking? My suspicion is that the network architecture might not be configured well for inference performance.

I can provide more specific guidance when I see what you're working with now, but some general guidance:

- If it's single animal and the crop size is not substantially smaller than the frame size, you're probably better off just running a **single instance** model. Top down will save time when the animals occupy a small proportion of the raw image size such that doing the crop saves a lot of computations in the second stage (centered instance model).
- Increasing the **output stride** of the model (to 2 or 4) will drastically reduce computation time at the cost of spatial resolution. (Though subpixel refinement might recover some of this, especially when using a sigma of 1.5-2.)
- If you can reduce the **max stride** to 8 or 16 without losing too much accuracy, you'll save a lot of computations since the higher max strides are really only beneficial when you have to cover a larger receptive field size.
- Reducing the **filters** and **filter rate** will decrease representational capacity but save on computations. Setting the filters to 16 or 24 and the filter rate to 1.5 will get you pretty lean architectures that are very fast.

If tweaking the model isn't possible without sacrificing accuracy on your data, the next thing to try is to [export the inference model](https://sleap.ai/guides/cli.html#sleap-export) as a `SavedModel` or `FrozenGraph` which can be converted to TFLite or TensorRT engines. Once you get to this step let me know and we can provide more guidance!

**Comment by lachioma** [comment]
Thank _you_ for your help!

Here is the `training_job_config.json` for the last model I trained: [training_config_files_lachioma.zip](https://github.com/talmolab/sleap/files/10361141/training_config_files_lachioma.zip).
I had found in [another discussion](https://github.com/talmolab/sleap/discussions/959#discussioncomment-3655760) suggestions from you on how to tune performance. So for this model, I already tried to increase output stride and reduce max stride, filter and filter rate.
I just ran inference on some frames and this is the inference latency I got:

![image](https://user-images.githubusercontent.com/29898879/211035796-2f40e2e5-114d-4184-98fe-cd91b18a8627.png)

I also tried to train a few other top-down models with slightly different parameters but the inference speed does not improve very much (in one case I got a latency of ~50 ms).

About using single_animal model, I tried to train a few of them. I am also attaching the `training_job_config.json` for one of them. For this, I get inference latency of 39+-5 ms.
![image](https://user-images.githubusercontent.com/29898879/211038556-4a935fc8-fb85-4e8d-b8e9-ca7b203abc48.png)
However, for all single_animal models I trained, I get the following warning when running inference:

`2023-01-06 10:00:52.382859: W tensorflow/core/data/root_dataset.cc:167] Optimization loop failed: Cancelled: Operation was cancelled
`
I searched on how to fix this, I tried to follow possible solutions indicated e.g. [here](https://discuss.tensorflow.org/t/optimization-loop-failed-cancelled-operation-was-cancelled/1524), but it didn't help.
